### PR TITLE
feat(build-tools): verify tool and skill references in skill Markdown

### PR DESCRIPTION
**Pull request by michaelva**
## Summary

Skill Markdown now carries explicit markup that the build resolves and verifies, so a documented tool reference cannot silently drift from the tool it describes. A reference that does not resolve, or an example a schema rejects, fails the build.

Five constructs, each rendering to what the model would have seen anyway — migrating a reference changes nothing at runtime, only whether drift is detectable:

| Written | Rendered | Verified at build time |
| --- | --- | --- |
| `{@tool fetch_document}` | `` `fetch_document` `` | the tool exists and the name is unambiguous |
| `{@skill web_search}` | `` `learn_web_search` `` | the skill exists and the name is unambiguous |
| `{@param search_documents.query}` | `` `query` `` | that tool's schema declares that field |
| `{@subagent_tool use_webbrowser browser_act}` | `` `browser_act` `` | both tools exist and both names are unambiguous |
| ` ```json tool=fetch_document ` | a plain ` ```json ` fence | the payload validates against that tool's schema |

`{@subagent_tool}` exists because a skill may legitimately document tools its own reader never gets — the tool it does get launches a second agent that has them. No graph can prove which actor a sentence addresses, so the author states it.

Checks fail closed throughout: an ambiguous name, a fence tagged for a tool whose schema this build cannot see, a `{@param …}` path crossing an unresolvable `$ref`, and constructs written inside code fences are all build errors rather than silent passes. Consumers point `vertesia-build.skillCatalog` at a module supplying their tool and skill names, and the Vite dev path loads the same catalog as the production build, so dev and CI cannot disagree.

Also here:

- `tools-sdk`: a `dispatch` descriptor, so a tool can declare which parameter carries another tool's name, and `supporting_tools`, so a skill can mark grants that need no instructions of their own.
- `common`: `PromptTemplate.parent` documented for what the runtime actually does with it — a fork is a draft that carries the prompt it was forked from, so `parent` is not evidence of a published record.
- Docs: README, IMPLEMENTATION and ASSETS-FEATURE rewritten for the current pipeline; unused `@rollup/*` dependencies dropped.

## Test Plan

- [x] `pnpm --filter @vertesia/build-tools test` — 177 passed
- [x] `pnpm --filter @vertesia/build-tools build` and `lint` — clean
- [x] Repo-wide `turbo build typecheck typecheck:test lint` in the consuming monorepo — 168/168 tasks

### CODE REVIEW

File: packages/build-tools/src/core/skill-markdown/preprocess.ts [hunk 1]
```diff
@@ -0,0 +1,522 @@
...
+/**
+ * The tool a heading names, when it names exactly one — `## batch_execute`, `` ## `batch_execute` ``
+ * or `## {@tool batch_execute}`.
+ *
+ * A section titled after a tool is the author stating, in prose, which tool the examples below
+ * belong to. That is independent evidence about the tag, and the only kind available: field overlap
+ * cannot tell two tools apart when both declare `id`.
+ */
+function toolNamedByHeading(line: string, tools: ReadonlySet<string>): string | undefined {
+    const heading = HEADING.exec(line);
```
[the remark this answers is not in the document]
Comment (michaelva) [reply packages/build-tools/src/core/skill-markdown/preprocess.ts:274]: Fixed in 1943d6ea.

File: packages/build-tools/src/core/skill-markdown/preprocess.ts [hunk 2]
```diff
@@ -0,0 +1,522 @@
...
+    if (!source.includes('{@')) {
+        return source;
+    }
+    const prefix = options.skillToolPrefix ?? DEFAULT_SKILL_TOOL_PREFIX;
+
+    const masked: string[] = [];
+    const text = source.replace(INLINE_CODE, (span) => {
+        masked.push(span);
+        return `${MASK_OPEN}${masked.length - 1}${MASK_CLOSE}`;
+    });
```
[the remark this answers is not in the document]
Comment (michaelva) [reply packages/build-tools/src/core/skill-markdown/preprocess.ts]: Fixed in 1943d6ea.

File: packages/build-tools/src/core/skill-markdown/preprocess.ts [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+    }
+
+    return rendered.replace(
+        new RegExp(`${MASK_OPEN}(\\d+)${MASK_CLOSE}`, 'g'),
+        (_, index: string) => masked[Number(index)],
+    );
+}
+
+function isClosingFence(line: string, marker: string): boolean {
+    const match = FENCE.exec(line);
```
[the remark this answers is not in the document]
Comment (michaelva) [reply packages/build-tools/src/core/skill-markdown/preprocess.ts:460]: Fixed in 1943d6ea.

File: packages/build-tools/src/core/skill-markdown/preprocess.ts [hunk 4]
```diff
@@ -0,0 +1,522 @@
...
+    const examples: SkillExample[] = [];
+    const errors: string[] = [];
+
+    // The tool named by the most recent heading, if it named one. Any other heading clears it, so a
+    // later unrelated section never inherits the claim.
+    let headingTool: string | undefined;
+    let headingLine = 0;
+
+    for (let i = 0; i < lines.length; ) {
+        const fence = FENCE.exec(lines[i]);
```
[the remark this answers is not in the document]
Comment (michaelva) [reply packages/build-tools/src/core/skill-markdown/preprocess.ts:486]: Fixed in 1943d6ea.

File: packages/build-tools/src/core/skill-markdown/preprocess.ts [hunk 5]
```diff
@@ -0,0 +1,522 @@
...
+        if (!fence) {
+            if (HEADING.test(lines[i])) {
```
[the remark this answers is not in the document]
Comment (michaelva) [reply packages/build-tools/src/core/skill-markdown/preprocess.ts:488]: Fixed in 1943d6ea.

**Comment by michaelva** [comment]
All five CodeQL alerts are addressed in 1943d6ea.

One of the three patterns was genuinely quadratic, and measurably so — the inline-code span, `(`+)(?:[^`]|(?!\1)`)*\1`, whose body can match a backtick two ways: 1.6ms at 2k backticks, 6.1ms at 4k, 25ms at 8k, i.e. fourfold per doubling. Skill sources never look like that, but this is a published library and its input is whatever a consumer passes, so the alert is worth clearing rather than dismissing.

It is now scanned instead of matched: each opening run is walked once and closed by the first equal-length run after it, which is what the backreference resolved to anyway. 64k backticks take 0.7ms.

The other two were theoretical but easy to remove: the heading's optional closing `#` run is trimmed in code rather than by a lazy group followed by two whitespace runs, and a fence's info string is anchored on a non-blank character so it cannot overlap the spaces before it. Behaviour is unchanged in both cases — four new tests pin the masking edge cases (unclosed run, a single backtick inside a double-backtick span, several spans on one line), and removing the unclosed-run branch fails one of them.

File: packages/build-tools/src/core/skill-markdown/preprocess.ts [hunk 6]
```diff
@@ -0,0 +1,570 @@
...
+/**
+ * The tool a heading names, when it names exactly one — `## batch_execute`, `` ## `batch_execute` ``
+ * or `## {@tool batch_execute}`.
+ *
+ * A section titled after a tool is the author stating, in prose, which tool the examples below
+ * belong to. That is independent evidence about the tag, and the only kind available: field overlap
+ * cannot tell two tools apart when both declare `id`.
+ */
+function toolNamedByHeading(line: string, tools: ReadonlySet<string>): string | undefined {
+    const heading = HEADING.exec(line);
```
[the remark this answers is not in the document]
Comment (michaelva) [reply packages/build-tools/src/core/skill-markdown/preprocess.ts:274]: Fixed in b5d11a1a. `HEADING` let `[ \t]+` and `[^\r\n]*` divide the same tabs, the same defect `FENCE` had; the text group is now anchored on a non-blank first character with an empty alternative, so backtracking the separator fails in O(1).

Measured on `"#" + "\t".repeat(30000) + "\r x"`: 999ms before, 0ms after. Behaviour is unchanged — the old and new patterns agree on match/no-match and on group 1 over 300k random inputs built from `#`, tabs, spaces, `\r` and text.

File: packages/build-tools/src/core/skill-markdown/preprocess.ts [hunk 7]
```diff
@@ -0,0 +1,570 @@
...
+            errors.push(
+                `line ${line}: '${whole}' is ambiguous — more than one provider defines the ${kind} ` +
+                    `'${name}'. Resolve the collision at the source before referencing it.`,
+            );
+        } else if (!known.has(name)) {
+            errors.push(`line ${line}: '${whole}' refers to a ${kind} no provider registers`);
+        }
+        references.push({ kind, name, rendered: `\`${target}\``, line, resolved });
+        return `\`${target}\``;
+    });
```
[the remark this answers is not in the document]
Comment (michaelva) [reply packages/build-tools/src/core/skill-markdown/preprocess.ts:445]: Fixed in b5d11a1a. `CONSTRUCT` let the keyword `[A-Za-z0-9_]*` and the argument group `[^}]*` both consume a word character, so `{@A0000…` with no closing brace had quadratically many ways to fail. The argument group is now `[^A-Za-z0-9_}][^}]*|` — a first character the keyword cannot take, or empty.

Measured on `"{@A" + "0".repeat(30000)`: 340ms before, 0ms after. Same 300k-input differential check as above: identical match offsets and both capture groups, including the `{@tool}` and `{@tolo x}` cases the reporting depends on.

File: packages/build-tools/src/core/skill-markdown/preprocess.ts [hunk 8]
```diff
@@ -0,0 +1,522 @@
...
+
+/** `{@word …}`. The keyword is matched loosely so `{@tolo x}` is a reported typo, not prose. */
+const CONSTRUCT = /\{@([A-Za-z][A-Za-z0-9_]*)([^}]*)\}/g;
+
+/**
+ * The same construct, restricted to the real keywords, for use inside fenced blocks. The loose
+ * form would flag template syntax that fenced examples legitimately contain — Handlebars
+ * `{{@index}}`, for one — which the renderer only ever sees in prose.
+ */
+const FENCED_CONSTRUCT = /\{@(?:tool|skill|param)\b[^}]*\}/g;
```
[the remark this answers is not in the document]
Comment (michaelva) [reply packages/build-tools/src/core/skill-markdown/preprocess.ts]: Correct, and fixed in b5d11a1a — `FENCED_CONSTRUCT` is now `/\{@(?:subagent_tool|tool|skill|param)\b[^}]*\}/g`.

This is the cost of the guard listing keywords explicitly rather than reusing the loose `CONSTRUCT` form, which it cannot: the loose form would flag Handlebars `{{@index}}` in a fenced example. So every new construct has to be added here too, and `subagent_tool` was not. A construct inside a fence is passed through unrendered and is not a reference either, so nothing else would have reported it — it would have reached the model as raw `{@subagent_tool …}`.

File: packages/build-tools/tests/skill-markdown.test.ts [hunk 9]
```diff
@@ -0,0 +1,736 @@
...
+    it('reports a construct written inside a code fence', () => {
+        const md = ['```yaml', '# see {@skill web_search} for details', '```'].join('\n');
+        const result = preprocessSkillMarkdown(md, { tools: new Set(), skills: new Set(['web_search']) });
+
+        expect(result.errors).toHaveLength(1);
+        expect(result.errors[0]).toContain('inside a code fence');
+        expect(result.errors[0]).toContain('line 2');
+        // Still passed through untouched: the fence body is never rewritten.
+        expect(result.markdown).toContain('{@skill web_search}');
+    });
```
[the remark this answers is not in the document]
Comment (michaelva) [reply packages/build-tools/tests/skill-markdown.test.ts:559]: Added in b5d11a1a — `reports a sub-agent construct written inside a code fence`, asserting the error, its line, and that the fence body is still passed through untouched.

Mutation-checked: dropping `subagent_tool` back out of `FENCED_CONSTRUCT` fails exactly that test and nothing else.

File: packages/build-tools/src/core/skill-markdown/preprocess.ts [hunk 10]
```diff
@@ -0,0 +1,522 @@
...
+ */
+function toolNamedByHeading(line: string, tools: ReadonlySet<string>): string | undefined {
+    const heading = HEADING.exec(line);
+    if (!heading) {
+        return undefined;
+    }
+    const named = heading[1]
+        .replace(/^\{@tool\s+([A-Za-z][A-Za-z0-9_]*)\}$/, '$1')
+        .replace(/^`([A-Za-z][A-Za-z0-9_]*)`$/, '$1');
+    return tools.has(named) ? named : undefined;
```
[the remark this answers is not in the document]
Comment (michaelva) [reply packages/build-tools/src/core/skill-markdown/preprocess.ts]: Correct, and fixed in b5d11a1a. `NAME` is `/^[a-z][a-z0-9_-]*$/`, so `web-fetch` is a legal reference, but the two unwrapping patterns in `toolNamedByHeading()` allowed only `[A-Za-z0-9_]`. The failure was silent in the worst way: the heading simply did not name a tool, so the tag-agreement check never ran under it and nothing was reported.

Both patterns now allow `-`, with a regression test covering `` ## `web-fetch` `` and `## {@tool web-fetch}`. Mutation-checked: reverting either pattern fails it.

File: packages/build-tools/src/core/skill-markdown/preprocess.ts [hunk 11]
```diff
@@ -0,0 +1,522 @@
...
+
+    const rendered = text.replace(CONSTRUCT, (whole, keyword: string, rest: string, offset: number) => {
+        consumed.add(offset);
+        const args = rest.trim().split(/\s+/).filter(Boolean);
+
+        if (keyword !== 'tool' && keyword !== 'skill' && keyword !== 'param' && keyword !== 'subagent_tool') {
+            errors.push(
+                `line ${line}: unknown construct '{@${keyword} …}' (expected {@tool …}, {@skill …}, ` +
+                    '{@param …} or {@subagent_tool …})',
+            );
```
[the remark this answers is not in the document]
Comment (michaelva) [reply packages/build-tools/src/core/skill-markdown/preprocess.ts:414]: Checked this one and it does not reproduce — no change made. The apostrophe pair closes around the construct, and the parenthesis closes the expected-list:

```
line 1: unknown construct '{@tolo …}' (expected {@tool …}, {@skill …}, {@param …} or {@subagent_tool …})
```

The `'` you flagged after `…})` is the JavaScript string delimiter of the second concatenated literal, not part of the message. Rendered by evaluating the two literals as written in the source.

File: packages/build-tools/src/core/transformers/skill.ts [hunk 12]
```diff
@@ -267,51 +276,106 @@ function buildSkillDefinition(
...
+    if (options) {
+        return assertSkillMarkdown(markdown, options, filePath).markdown;
+    }
+    const probe = preprocessSkillMarkdown(markdown, { tools: new Set(), skills: new Set() });
+    if (probe.references.length > 0 || probe.examples.length > 0 || probe.errors.length > 0) {
+        throw new Error(
+            `${filePath} uses {@tool …}, {@skill …} or a tool= example tag, but this build supplies no ` +
+                'skill catalog. Set "vertesia-build.skillCatalog" in package.json to a module exporting ' +
+                '{ tools, skills } (and optionally validateExample).',
+        );
```
[the remark this answers is not in the document]
Comment (michaelva) [reply packages/build-tools/src/core/transformers/skill.ts:305]: Checked this one too and it does not reproduce — no change made. The backtick you flagged is the template literal's own closing delimiter, immediately before `+`; the concatenation is clean:

```
a.md uses {@tool …}, {@skill …} or a tool= example tag, but this build supplies no skill catalog. Set "vertesia-build.skillCatalog" in package.json to a module exporting { tools, skills } (and optionally validateExample).
```

Rendered by evaluating the three literals as written in the source.

**Comment by michaelva** [comment]
Pushed `b5d11a1a`. It closes the two new CodeQL alerts (248/249) and the three Copilot findings that reproduced, and — the important one — a superlinear scan that the earlier CodeQL fix introduced.

**The masking scan was still quadratic.** Replacing the backreference regex removed the engine's backtracking but not the algorithm's cost: an opening backtick run with no closer called `indexOf` over the whole remainder, so runs in *descending* length made all of them sweep to the end. `Array.from({length: 1200}, (_, i) => '`'.repeat(1200 - i)).join('x')` — 722 KB — took **18.3s**. The comment claiming each run is walked once was wrong.

It now tokenises the backtick runs once and precomputes a suffix maximum, so "is any later run long enough to close this one?" is O(1). Openers that fail cost nothing; openers that succeed consume everything they scanned past. Same input: **2.1ms**. Output is byte-identical to the previous implementation — checked over 40k random inputs of backtick runs and filler, 0 mismatches.

Two things worth recording:

- The regression test at `skill-markdown.test.ts` was toothless for a reason neither of us named: `renderLine()` returns early unless the line contains `{@`, so `'`'.repeat(20000)` never reached the scanner at all. That also bounds the exposure — reaching this needs a construct on the same line — but it is a published library and the function is superlinear regardless. Both timing tests now carry a construct so they actually exercise it, and the descending-runs case is one of them. Mutation-checked: restoring the old scanner fails it at 18.34s against the 1s threshold.
- The two remaining CodeQL alerts were the same defect in two more patterns: `HEADING` let its separator and its text both match a tab; `CONSTRUCT` let its keyword and its argument group both match a word character. Both are now anchored the way `FENCE` already was. 30k-character adversarial inputs: 999ms → 0ms and 340ms → 0ms, with match results identical over 300k random inputs.

Two Copilot findings did not reproduce — the "stray apostrophe" and "stray backtick" are the JavaScript string delimiters of the concatenated literals, not message content. Rendered output is in the thread replies.

Verification: `build-tools` 185/185 (was 181, +4), `biome check src tests` clean, `tsc` build clean. Downstream, the skill sources rebuild and `check:skills` exits 0 with no new findings.

File: packages/build-tools/src/core/skill-markdown/preprocess.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (michaelva) [reply packages/build-tools/src/core/skill-markdown/preprocess.ts:445]: Reopened, and you were right to leave it open — my first fix addressed the wrong half. Fixed properly in `4847732e`.

The new witness (`{@A/` repeated) is not backtracking inside one match; it is the **scan**. `[^}]*` could match `{`, so every `{@` on the line was a candidate start and each one swept to the end of the string before failing. Anchoring the argument group only removed the intra-match ambiguity, which is why the alert came back with a different pump.

A construct's arguments are names and separators — they never contain `{` — so excluding it bounds each attempt by the distance to the next candidate start:

| line | before | after |
| --- | --- | --- |
| 80 KB (20k openers) | 578ms | 1ms |
| 160 KB (40k openers) | 2296ms | 1ms |
| 240 KB (60k openers) | 5205ms | 1ms |

Fourfold per doubling before, flat after. `FENCED_CONSTRUCT` took the same class — it had the identical shape and would have been the next alert — and that also makes the two agree: neither now matches a construct with a brace in its arguments, where before the in-fence guard reported something the renderer would not have rendered.

The regression test is sized at 40k openers so the previous form fails it at 2.3s against a 1s threshold, rather than passing at 578ms the way a 20k-opener test would have. Mutation-checked both ways. Behaviour is otherwise unchanged: over 400k random inputs the only differences are constructs containing `{` in their arguments, which are malformed in prose too.

`build-tools` 186/186, `biome check` clean, and downstream `check:skills` still exits 0 with no new findings.

**Review by LeonRuggiero — APPROVED** [review]
